File: backend/lms/src/main/java/com/el/enrollment/application/dto/EnrolmentWithCourseDTO.java
package com.el.enrollment.application.dto;

import com.el.course.domain.Course;
import com.el.course.domain.CourseSection;
import com.el.course.domain.Language;
import com.el.enrollment.domain.*;

import java.time.LocalDateTime;
import java.util.Set;

/**
 * Combination DTO class between {@link CourseEnrollment} and {@link Course}.
 * */
public record EnrolmentWithCourseDTO(
        Long courseId,
        String title,
        String thumbnailUrl,
        String description,
        Language language,
        Set<Language> subtitles,
        Set<String> benefits,
        Set<String> prerequisites,
        Set<CourseSection> sections,
        String teacher,
        Long enrollmentId,
        String student,
        Set<LessonProgress> lessonProgresses,
        Boolean completed,
        LocalDateTime completedDate,
        Certificate certificate,
        Progress progress,
        Set<QuizSubmission> quizSubmissions
) {
    public static EnrolmentWithCourseDTO of(CourseEnrollment enrolment, Course course) {
        return new EnrolmentWithCourseDTO(
                // Course fields
                course.getId(),
                course.getTitle(),
                course.getThumbnailUrl(),
                course.getDescription(),
                course.getLanguage(),
                course.getSubtitles(),
                course.getBenefits(),
                course.getPrerequisites(),
                course.getSectionForPublished(),
                course.getTeacher(),
                // Enrolment fields
                enrolment.getId(),
                enrolment.getStudent(),
                enrolment.getLessonProgresses(),
                enrolment.getCompleted(),
                enrolment.getCompletedDate(),
                enrolment.getCertificate(),
                enrolment.getProgress(),
                enrolment.getQuizSubmissions()
        );
    }
}

File: angular-ui/src/app/enrolments/model/enrolment-with-course-dto.ts
import {LessonProgress} from "./lesson-progress";
import {Section} from "../../administration/courses/model/view/section";

export interface Progress {
  totalLessons: number,
  completedLessons: number,
}

export interface Certificate {
  id: number,
  fullName: string,
  email: string,
  student: string,
  teacher: string,
  url: string,
  courseId: number,
  courseTitle: string,
  issuedDate: string,
  certified: boolean
}

export interface EnrolmentWithCourseDto {
  courseId: number,
  title: string,
  thumbnailUrl: string,
  description: string,
  language: string,
  subtitles: string[],
  benefits: string[],
  prerequisites: string[],
  sections: Section[],
  teacher: string,
  enrollmentId: number,
  student: string,
  lessonProgresses: LessonProgress[],
  completed: boolean,
  completedDate: string,
  certificate: Certificate,
  progress: Progress
}
